### PR TITLE
Maintain ABI compatibility between release and debug builds

### PR DESCRIPTION
**Pull request by mpercy**
Summary: Because the integrations we are currently developing don't
maintain a well defined public API boundary with Kudu, adding fields to
structs in debug mode is unfortunately dangerous and crash prone. At
least for now, we need the ability to build Kudu in release mode and run
linked modules (Raft callbacks, etc) in debug mode. Those modules may
end up #including all kinds of headers (locks, etc) and
difficult-to-understand runtime crashes are the result.

In a perfect world, we would define an ABI-compatible interface inside
of KuduRaft and prevent exposing details of the locks we are using
inside of Log, RaftConsensus, etc. In the current world, we want the
flexibility to hack through and get something working quickly without
having to build in release mode only.

This patch disables several debug-mode-only features in Kudu (keeping
track of mutex owners, etc) because they end up changing the structs
defined in commonly-used header files. This fixes some crashes we've
been seeing when integrating some plugin code.

Test Plan: I ran a gtest that reliably reproduced a nasty crash and now
it doesn't crash anymore.

Reviewers: anirbanr-fb, abhinav04sharma

### CODE REVIEW

File: src/kudu/gutil/ref_counted.cc
Comment (anirbanr-fb) [inline src/kudu/gutil/ref_counted.cc]: in optimized mode of tp2 compilation, we use -DNDEBUG. In DEBUG mode, we don't
This would bring in the in_dtor_ extra member in DEBUG mode making it incompatible with OPT mode.
NDEBUG means "NO DEBUG"
NDEBUG_DISABLED_FB - would mean "NO DEBUG IS DISABLED IN FB"
To me seems like a double negative. Shouldn't it be named NDEBUG_FB?

Comment (mpercy) [reply src/kudu/gutil/ref_counted.cc]: It's even worse than that! `#ifndef NDEBUG` is already a double negative: "enable if not debug is not defined" . So this would actually be a triple negative: "enable if not debug is disabled in FB is not defined" 😜

My intention is simply to disable this #ifdef without deleting the lines (to minimize the possibility of merge conflicts) so I thought appending _DISABLED_FB to the existing `#ifndef NDEBUG` lines would be a good way to do that.

Maybe I should just do:

```
#ifndef FB_DO_NOT_REMOVE // NDEBUG
```

What do you think?

Comment (anirbanr-fb) [reply src/kudu/gutil/ref_counted.cc]: We can use FB_DO_NOT_REMOVE. That way we have one define to find out all code that we have touched.
Its also unlikely that we will use the special checks or assertions around in_dtor_

**Comment by mpercy** [comment]
Duh, `FB_DO_NOT_REMOVE` does not work because these are `#ifndef` statements so I changed it to `NDEBUG_FB_DO_NOT_REMOVE`. Seems good enough to me.

**Review by anirbanr-fb — APPROVED** [review]

**Comment by anirbanr-fb** [comment]
you can merge the PR when you are done. I will probably add a couple of PR before we cut the next TP2

**Comment by mpercy** [comment]
I went with: `#ifdef FB_DO_NOT_REMOVE  // #ifndef NDEBUG`

Please go ahead and merge this PR or add me to the write access ACL so I can.

**Comment by mpercy** [comment]
Actually, do not merge this. there is an issue in debug mode I'm working on fixing now.

**Comment by mpercy** [comment]
OK, go ahead and merge this please.

**Pull request merged**
